### PR TITLE
Add error handling for new site creation failure modes

### LINKED ISSUE #41067: Create site loading screen does not handle errors
If an error occurs while creating the site, the loading animation will stay stuck on the Applying design step.
The user should be notified of the error and ideally have a way to try again.
Screen Shot 2020-04-14 at 8 03 46 AM

#### Steps to reproduce
One way to simulate an error creating the site is to disable the network connection whil on the loading page:

In Chrome browser:
Fill in necessary gutenboarding details to get up to the https://wordpress.com/new/style page.
Open the dev tools *network* tab, see where it says *Online ˅* 
Click on Continue, land on the loading page
Quickly change *Online ˅* to *Offline* before the editor loads!

#### What I expected
Should see an error message, and an option to try again

#### What happened instead
Stuck on the loading screen :(

### PR DESCRIPTION
**Pull request by lsl**
#### Changes proposed in this Pull Request

* Displays an error message if create site returns with an error.

#### Testing instructions

Full flow:
* Apply D44621-code
* Attempt to create a site using http://calypso.localhost:3000/new
* Clicking `Continue` on the `/new/plans` step should begin showing `<CreateSite />` before transitioning to the error message showing a "Go Back" button to return to the last known good step.

Error only:
* `ENTRY_LIMIT=entry-gutenboarding yarn start`
* Visit http://calypso.localhost:3000/new/error (I'll remove this route before merge)

Mobile:
Peek 2020-06-15 13-57-mobile

Desktop:
Peek 2020-06-15 13-54

Fixes #41067

### CODE REVIEW

**Comment by dubielzyk** [comment]
>Apply D44621-code

I'm not familiar how to do this to test it out?

> Styling (nothing defined yet)

Here's a mock of how it should look like:
**Desktop**
image

**Mobile**
image

You can check out the mocks in Figma.

Do you want me to review it now or when the styling is done @lsl ?

**Comment by lsl** [comment]
Those mocks are great, thanks @dubielzyk.

I'll add a route to this PR for you to load the error without needing the help of anything serverside.

> now or when the styling is done

I'll ping you again once it's cleaned up. Thanks!

**Comment by lsl** [comment]
Those mocks are great, thanks @dubielzyk.

I'll add a route to this PR for you to load the error without needing the help of anything serverside.

> now or when the styling is done

I'll ping you again once it's cleaned up. Thanks!

**Comment by lsl** [comment]
Hey @dubielzyk, would you be able to take another look at this?

There's a temp route to display the error at: http://calypso.localhost:3000/new/error

**Comment by lsl** [comment]
@andrewserong I did a rebase to pull out the temp route easier later but I think I broke your changes requested. I think I got everything discussed though, are you able to take a look at this again?

**Comment by andrewserong** [comment]
This is looking really good @lsl! The only change I'd make now is to remove the 'Contact Support' button, and add a link to contact support within the paragraph, so that we can match the text in the designs that Jon added. We can use `createInterpolateElement` to swap out the tag placeholder for an `ExternalLink` component. [Here's an example](https://github.com/automattic/wp-calypso/blob/a4e7811614eec7908116449c8371cfa861b61dc3/client/landing/gutenboarding/components/signup-form/index.tsx#L112) from the terms of service in the signup form.

The other thing to watch out for using the Link component is that it resolves to a `react-router-dom` `<RouterLink>` which tries to redirect within Gutenboarding instead of going to a full URL. So the button currently goes to `/new/https://wordpress.com/support/contact`. For external links, we've mostly been using `ExternalLink` so far (it adds in a little icon to indicate it'll open in a new window, too).

File: client/landing/gutenboarding/components/create-site-error/style.scss [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	right: 0;
+	bottom: 0;
+	z-index: 50;
+	background-color: $white;
+
+	.create-site-error__header {
+		display: flex;
+		justify-content: space-between;
+		align-items: center;
+		height: 56px;
```
Comment (andrewserong) [inline client/landing/gutenboarding/components/create-site-error/style.scss:17]: I think we should be able to get the vertical position of the `W` logo to line up if we set this to `64px` to match the other steps in the flow.

**Comment by andrewserong** [comment]
Just added the "[Status] Needs Review" label to kick off the e2e tests.

**Comment by andrewserong** [comment]
Actually, I might be off with those CSS values, I think it's changed a little bit on `master`, so it might be clearer after a rebase.

**Comment by lsl** [comment]
@andrewserong The rebase broke a couple of things, I've just pushed a couple of WIP commits to show you what I'm thinking about doing instead. your suggestion to remove the overlay setup and use the header logic seems to work well. 

I've made a few extra changes to tidy things up and make the transitions smoother, keeping the titles vert aligned (on mobile at least)

The `<CreateSite>` is where I got the html for the error for so I figure we should probably update that while we're at it.

Although I did note it did not have the same header problem:
Screenshot_2020-06-15 WordPress com

Also, for consistency sake I made use of `<Title>` and `<ActionButtons>` components similar to the other pages. @dubielzyk this could have made things better or worse, I'm not confident in deciding which is better for usability, could you let me know if I should keep the "Go back" button close to the error message on mobile or have it down in the Action button area like on other pages:

Screenshot_2020-06-15 WordPress com(1)

File: client/landing/gutenboarding/onboarding-block/edit.tsx [hunk 2]
```diff
@@ -89,6 +91,16 @@ const OnboardingEdit: FunctionComponent< BlockEditProps< Attributes > > = () =>
 
 	const redirectToLatestStep = <Redirect to={ getLatestStepPath() } />;
 
+	function createSiteOrError() {
+		if ( newSiteError ) {
```
Comment (ramonjd) [inline client/landing/gutenboarding/onboarding-block/edit.tsx:95]: `newSiteError` is still present when I return to the plans page from the error page.

That means if I navigate back using the browser button, I'll hit the error:

Jun-15-2020 13-08-31

It should it the `redirectToLatestStep`.

File: client/landing/gutenboarding/components/create-site-error/index.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import { localizeUrl } from 'lib/i18n-utils';
+import ActionButtons from '../../components/action-buttons';
+import { Title, SubTitle } from '../../components/titles';
+interface Props {
+	linkTo: string;
+}
+
+const CreateSiteError: FunctionComponent< Props > = ( { linkTo } ) => {
+	const { __ } = useI18n();
+
```
Comment (ramonjd) [inline client/landing/gutenboarding/components/create-site-error/index.tsx:24]: To address the [browser navigation point below](https://github.com/Automattic/wp-calypso/pull/43092/files#r439910944), I think we might need to clear the site store here when the component unmounts, e.g., 

```
	const { reset: resetSignupStore } = useDispatch( SITE_STORE );

	useEffect( () => resetErrors );

```

It should be okay to clear the entire site store, but we could also create an action to just reset the site errors in the site store actions file:

https://github.com/Automattic/wp-calypso/blob/bb8372d/packages/data-stores/src/site/actions.ts

What do you think?

File: client/landing/gutenboarding/onboarding-block/edit.tsx (hunk 2 above)
Comment (lsl) [reply client/landing/gutenboarding/onboarding-block/edit.tsx:95]: I'm unable to replicate in firefox, I do have an unresponsive back button though. I'll look at resetting the error state after routing to the error.

File: client/landing/gutenboarding/onboarding-block/create-site/style.scss [hunk 4]
```diff
@@ -1,83 +1,54 @@
...
 
-.create-site__background {
-	position: fixed;
-	top: 0;
-	left: 0;
-	right: 0;
-	bottom: 0;
-	z-index: 50;
-	background-color: $white;
+.create-site {
```
Comment (lsl) [inline client/landing/gutenboarding/onboarding-block/create-site/style.scss:10]: Diff for this css is pretty hard to understand, basically I've flattened the html and shifted the nesting here up two? levels, very little css was changed.

File: client/landing/gutenboarding/onboarding-block/edit.tsx (hunk 2 above)
Comment (ramonjd) [reply client/landing/gutenboarding/onboarding-block/edit.tsx:95]: Hmmm. I do not rule out that it's just me :) 7/10 it is.

File: client/landing/gutenboarding/components/create-site-error/index.tsx (hunk 3 above)
Comment (lsl) [reply client/landing/gutenboarding/components/create-site-error/index.tsx]: So this scenario:

/new/plans -> /new/create-site -> /new/create-site (error) -> click "Go Back" -> /new/plans -> click browser back button -> /new/create-site (error)

What was happening is the "Go back" wasn't a go back but a link to the plans page, hitting browser back would take you back to the error message. I think this actually makes a bit of sense, if you hit back again you'll end up back at the plans page.

I don't know if this is something that needs fixing (da3c356 has an attempt) and what is the desired outcome be if you select back while on the plans page after seeing an error?

File: packages/data-stores/src/site/reducer.ts [hunk 5]
```diff
@@ -30,6 +30,7 @@ export const newSiteError: Reducer< NewSiteErrorResponse | undefined, Action > =
 		case 'FETCH_NEW_SITE':
 		case 'RECEIVE_NEW_SITE':
 		case 'RESET_SITE_STORE':
+		case 'RESET_RECEIVE_NEW_SITE_FAILED':
```
Comment (ramonjd) [inline packages/data-stores/src/site/reducer.ts:33]: 👍

File: client/landing/gutenboarding/onboarding-block/edit.tsx [hunk 6]
```diff
@@ -126,8 +138,11 @@ const OnboardingEdit: FunctionComponent< BlockEditProps< Attributes > > = () =>
 					<Plans isModal />
 				</Route>
 
-				<Route path={ makePath( Step.CreateSite ) }>
-					{ canUseCreateSiteStep() ? <CreateSite /> : redirectToLatestStep }
+				<Route path={ makePath( Step.CreateSite ) }>{ createSiteOrError() }</Route>
+
+				{ /* tofix: remove this testing route (and step def in path.ts) */ }
```
Comment (ramonjd) [inline client/landing/gutenboarding/onboarding-block/edit.tsx:143]: Are you still removing this?

File: client/landing/gutenboarding/onboarding-block/create-site/style.scss [hunk 7]
```diff
@@ -1,83 +1,54 @@
...
-		display: flex;
-		justify-content: space-between;
-		align-items: center;
-		height: 56px;
-		padding: 8px;
-	}
-
-	.create-site__content {
-		position: absolute;
-		top: calc( 50% - #{$progress-widget-height / 4} );
```
Comment (andrewserong) [inline client/landing/gutenboarding/onboarding-block/create-site/style.scss:32]: I think removing this container might be causing the progress bar to no longer be centred vertically on mobile.

On production, the mobile layout of the create-site step looks like:

image

With this change, it's moved to the top:

image

**Comment by andrewserong** [comment]
The error message page is looking great to me now! Just left a comment about the vertical position of the progress bar on mobile.

File: client/landing/gutenboarding/onboarding-block/create-site-error/index.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import React, { FunctionComponent, useEffect } from 'react';
+import { useI18n } from '@automattic/react-i18n';
+import { ExternalLink } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import Link from '../../components/link';
+import './style.scss';
+import { createInterpolateElement } from '@wordpress/element';
```
Comment (andrewserong) [inline client/landing/gutenboarding/onboarding-block/create-site-error/index.tsx:13]: Tiniest of nits: these `@wordpress` imports should be in the `External dependencies` section above.

**Comment by lsl** [comment]
Thanks @andrewserong, cleaned up those dep comments, rebased, retested. Looking good to go at this point. :crossed_fingers:

**Comment by andrewserong** [comment]
Nice @lsl — just gave it another test, and still looking good to me. On the `create-site` page, In the happy path where an error isn't thrown, now that we're showing the header, there's a tiny moment when the site title reverts back to "Start your website" before the redirect happens (after the store is cleared and before the redirect).

creating-site-small

I don't think it's a blocker, but if it does come up, we can fix it up in a follow up.

**Comment by lsl** [comment]
Thanks @andrewserong! Good catch, thoughts on fde709e as a workaround?

**Comment by andrewserong** [comment]
Nice workaround @lsl — thanks for adding the comment there, too for the next person working on it! 👍

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3860992).

File: client/landing/gutenboarding/onboarding-block/create-site-error/index.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+const CreateSiteError: FunctionComponent< Props > = ( { linkTo } ) => {
+	const { __ } = useI18n();
+
+	const { resetNewSiteFailed } = useDispatch( SITE_STORE );
+
+	useEffect( () => resetNewSiteFailed );
+
+	return (
+		<div className="create-site-error">
+			<Title>{ __( 'Something went wrong…' ) }</Title>
```
Comment (a8ci18n) [inline client/landing/gutenboarding/onboarding-block/create-site-error/index.tsx:36]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **51** times:
`translate( 'Oops! Something went wrong.' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/create-site-error)

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3860992).

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3860992).

**Pull request merged**